src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -70,7 +70,72 @@ public sealed partial class JsonSerializerOptions
             return converters;
 
             void Add(JsonConverter converter) =>
-                converters.Add(converter.TypeToConvert!, converter);
+                converters.Add(converter.TypeToConvert, converter);
+        }
+
+        internal JsonConverter GetDictionaryKeyConverter(Type keyType)
+        {
+            if (_dictionaryKeyConverters == null)
+            {
+                _dictionaryKeyConverters = GetDictionaryKeyConverters();
+            }
+
+            if (!_dictionaryKeyConverters.TryGetValue(keyType, out JsonConverter? converter)){
+                if (keyType.IsEnum)
+                {
+                    converter = GetEnumConverter();
+                    _dictionaryKeyConverters[keyType] = converter;
+                }
+                else
+                {
+                    ThrowHelper.ThrowNotSupportedException_DictionaryKeyTypeNotSupported(keyType);
+                }
+            }
+
+            return converter!;
+
+            JsonConverter GetEnumConverter()
+                => (JsonConverter)Activator.CreateInstance(
+                        typeof(EnumConverter<>).MakeGenericType(keyType),
+                        BindingFlags.Instance | BindingFlags.Public,
+                        binder: null,
+                        new object[] { EnumConverterOptions.AllowStrings | EnumConverterOptions.AllowNumbers, this },
+                        culture: null)!;
+        }
+
+        private ConcurrentDictionary<Type, JsonConverter>? _dictionaryKeyConverters;
+
+        private static ConcurrentDictionary<Type, JsonConverter> GetDictionaryKeyConverters()
+        {
+            const int NumberOfConverters = 18;
+            var converters = new ConcurrentDictionary<Type, JsonConverter>(Environment.ProcessorCount, NumberOfConverters);
+
+            // When adding to this, update NumberOfConverters above.
+            Add(new BooleanConverter());
```

Comment:
It seems like these could use the same instance of the converters that we already instantiate.

To do that, we could add a field for each default converter on the options class (and use that in both cases -- here and when initializing `s_defaultSimpleConverters`), or to avoid the extra cost of the field storage, instead loop through the existing built-in converters (`s_defaultSimpleConverters`) and add the known ones perhaps by calling a new internal virtual method "internal virtual bool SupportsQuotedNumbers {get}".

Reply:
I like the idea of re-use the instances from `s_defaultSimpleConverters`. Is there a benefit on doing so? I think we would only lower warm-up cost.

Reply:
Warm-up cost would be less and of course RAM would be a bit less since there are fewer objects in memory. 

Plus there would be a few less references to investigate during a GC (eventually these will be in gen 3 pass so not a big deal since they will happen infrequently)